src/delta.c

```diff
@@ -38,86 +38,159 @@
+end:
+	free_string(&prefixed_from_file);
 }
 
-static void do_delta(char *filename, struct file *file)
+/* Check if the delta filename is well-formed, if so return true and fill the from/to
+ * buffers with the corresponding hashes. Return false otherwise. */
+static bool check_delta_filename(const char *delta_name, char *from, char *to)
 {
-	char *origin;
-	char *dir, *base, *tmp = NULL, *tmp2 = NULL;
-	char *deltafile = NULL;
-	int ret;
-	struct stat sb;
+	/* Delta files have the form [FROM_VERSION]-[TO_VERSION]-[FROM_HASH]-[TO_HASH]. */
+	const char *s = delta_name;
+
+	/* Ignore versions, deltas will be used based on their hashes only. */
+	for (int i = 0; i < 2; i++) {
+		s = strchr(s, '-');
+		if (!s) {
+			return false;
+		}
+		s++;
+	}
 
-	string_or_die(&deltafile, "%s/delta/%i-%i-%s-%s", state_dir,
-		      file->deltapeer->last_change, file->last_change, file->deltapeer->hash, file->hash);
+	/* Note: SWUPD_HASH_LEN accounts for the NUL-terminator after the hash. */
+	const size_t hash_len = SWUPD_HASH_LEN - 1;
+	if (strlen(s) != (hash_len * 2 + 1)) {
+		return false;
+	}
 
-	ret = stat(deltafile, &sb);
-	if (ret != 0) {
-		free_string(&deltafile);
-		return;
+	if (s[hash_len] != '-') {
+		return false;
 	}
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
+	memcpy(from, s, hash_len);
 
-	dir = dirname(tmp);
-	base = basename(tmp2);
+	/* Consume the first hash and the separator. */
+	s += hash_len + 1;
 
-	string_or_die(&origin, "%s/%s/%s", path_prefix, dir, base);
+	memcpy(to, s, hash_len);
 
-	ret = apply_bsdiff_delta(origin, filename, deltafile);
-	if (ret) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
-	xattrs_copy(origin, filename);
+	return true;
+}
 
-	if (!verify_file(file, filename)) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
+void apply_deltas(struct manifest *current_manifest)
+{
+	char *delta_dir;
+	string_or_die(&delta_dir, "%s/delta", state_dir);
 
-	if (xattrs_compare(origin, filename) != 0) {
-		unlink_all_staged_content(file);
-		goto out;
+	DIR *dir = opendir(delta_dir);
+	if (!dir) {
+		/* No deltas available to apply. */
+		return;
 	}
 
-	unlink(deltafile);
+	struct dirent *ent;
+	while ((ent = readdir(dir))) {
+		if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, "..")) {
+			continue;
+		}
+
+		char *to_staged = NULL;
+
+		char *delta_name = ent->d_name;
+		char *delta_file;
+		string_or_die(&delta_file, "%s/%s", delta_dir, delta_name);
+
+		char from[SWUPD_HASH_LEN] = {};
+		char to[SWUPD_HASH_LEN] = {};
+		if (!check_delta_filename(delta_name, &from[0], &to[0])) {
+			fprintf(stderr, "Invalid name for delta file: %s\n", delta_file);
+			goto next;
+		}
+
+		string_or_die(&to_staged, "%s/staged/%s", state_dir, to);
+
+		/* If 'to' file already exists, no need to apply delta. */
+		struct stat stat;
+		if (lstat(to_staged, &stat) == 0) {
+			goto next;
+		}
+
+		/* TODO: Sort the list of files by hash then walk it in parallel with readdir. */
+
```

Comment:
This will reduce the number of hash compares we do a lot.

Reply:
Agreed (hence the TODO). I've opted to get the version we have in first, as it introduces a change in how we pick the deltas. Once that is in, the walk optimization should be a straightforward patch.

Reply:
This is an important TODO. Not going to block the merge, just noting that the TODO is important...

Reply:
From the `readdir` man page: 

> The order in which filenames are read by successive calls to readdir() depends on the 
> filesystem implementation; it is unlikely that the names will be sorted in any fashion.

So how is this a solution? Maybe `scandir` with `alphasort` (which I found after some googling).

Reply:
Looking at a typical profile, delta application doesn't seem to be what is causing such difference, as the time it takes is lower than the difference you are seeing. I suspect there are some other noise factors at play. If we were to optimize, we should look to improve the consolidation functions first.

I'll go back to my initial position: keep this TODO (I've updated to take into account the ordering), and we later can make this improvement. I'll see if we have some low hanging fruit in the consolidation instead (independent of this PR).

```
Statistics for going from version 21810 to version 21820:

    changed bundles   : 14
    new bundles       : 0
    deleted bundles   : 0

    changed files     : 12
    new files         : 2
    deleted files     : 2

Starting download of remaining update content. This may take a while...
	...100%
Finishing download of update content...
Staging file content
Applying update
	...100%
Update was applied.
Calling post-update helper scripts.
Update took 5.1 seconds

Raw elapsed time stats:
    144.57 ms: Main Update
     12.71 ms: Update Step 1: get versions
     94.40 ms: Load Manifests:
   4262.79 ms: Recurse and Consolidate Manifests
   1588.24 ms: Add Included Manifests
      0.54 ms: Pre-Update Scripts
    177.92 ms: Download Packs
     38.25 ms: Apply deltas
      5.40 ms: Create Update List
    228.30 ms: Update Loop
      0.02 ms: Run Scripts

CPU process time stats:
      3.32 ms: Main Update
      0.21 ms: Update Step 1: get versions
      1.48 ms: Load Manifests:
   2393.18 ms: Recurse and Consolidate Manifests
    477.46 ms: Add Included Manifests
      0.54 ms: Pre-Update Scripts
     13.37 ms: Download Packs
     38.06 ms: Apply deltas
      5.38 ms: Create Update List
     44.33 ms: Update Loop
      0.02 ms: Run Scripts
Update successful. System updated from version 21810 to version 21820
```

Reply:
Thanks for your additional measurements. I wasn't actually too concerned with the time differential, as I said "With the variation I saw in just doing 10 updates I believe this is insignificant. The TODO can wait until after this PR is merged."

Reply:
Yes, I got that. I was just thinking that we had a bigger problem in our hands, hence the additional measurements :-)